File: develop/client/game/Assets/src/game/scene/unit/GUnitPosLogic.cs
using System;
using System.Diagnostics;
using ShineEngine;
using UnityEngine;
using UnityEngine.UI;

/// <summary>
/// 单位位移控制逻辑
/// </summary>
public class GUnitPosLogic:UnitPosLogic
{
	/** 游戏单位 */
	private GUnit _gUnit;

	private GUnitPosData _gd;

	private DirData _tShootDir=new DirData();

	private int _sendShootDirTime=0;

	public GUnitPosLogic()
	{

	}

	public override void init()
	{
		base.init();

		_gUnit=(GUnit)_unit;
		_gd=(GUnitPosData)_data.pos;
	}

	public override void dispose()
	{
		base.dispose();

		_gUnit=null;
		_gd=null;
	}

	public override void onFrame(int delay)
	{
		base.onFrame(delay);

		if(_sendShootDirTime>0)
		{
			if((_sendShootDirTime-=delay)<=0)
			{
				_sendShootDirTime=0;
				CUnitChangeShootDirRequest.create(_unit.instanceID,_gd.shootDir).send();
			}
		}
	}

	public void onSetShootDir()
	{
		((GUnitShowLogic)_unit.show).setShootDir(_gd.shootDir);
	}

	public void setShootDirByServer(DirData data)
	{
		_gd.shootDir=data;
		onSetShootDir();
	}

	public void setShootDir(float dir)
	{
		bool changed;

		if(_gd.shootDir==null)
		{
			_gd.shootDir=_tShootDir;

			changed=true;
		}
		else
		{
			changed=_gd.shootDir.direction!=dir;
		}

		if(changed)
		{
			_gd.shootDir.direction=dir;

			onSetShootDir();

			doSendShootDir();
		}
	}

	private void doSendShootDir()
	{
		if(_sendShootDirTime==0)
		{
			CUnitChangeShootDirRequest.create(_unit.instanceID,_gd.shootDir).send();
			_sendShootDirTime=200;
		}
	}

	public void cancelShootDir()
	{
		_gd.shootDir=null;
		onSetShootDir();
		CUnitChangeShootDirRequest.create(_unit.instanceID,null).send();
		_sendShootDirTime=0;
	}
}

File: develop/server/project/sceneBase/src/main/java/com/home/sceneBase/scene/unit/GUnitPosLogic.java
package com.home.sceneBase.scene.unit;

import com.home.base.data.scene.unit.GUnitPosData;
import com.home.commonBase.data.scene.base.DirData;
import com.home.commonSceneBase.scene.unit.BUnitPosLogic;
import com.home.sceneBase.net.sceneBaseRequest.unit.UnitChangeShootDirRequest;

public class GUnitPosLogic extends BUnitPosLogic
{
	private GUnitPosData _gd;
	
	@Override
	public void init()
	{
		super.init();
		
		_gd=(GUnitPosData)_d;
	}
	
	@Override
	public void dispose()
	{
		super.dispose();
		
		_gd=null;
	}
	
	public void onChangeShootDir(DirData data)
	{
		if(data!=null)
		{
			if(_gd.shootDir==null)
				_gd.shootDir=data;
			else
				_gd.shootDir.copyDir(data);
		}
		else
		{
			_gd.shootDir=null;
		}
		
		_unit.radioMessage(UnitChangeShootDirRequest.create(_unit.instanceID,data),false);
	}
}
